<a href="https://colab.research.google.com/github/QY428/Machine-Learning-Practice/blob/main/Code/Ex3_Keras_for_Beginners_Building_Your_First_Neural_Network.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Follow the above tutorial and try to build your first neural network and upload your ipynb to Ummoodle showing the accuracy rate of your final result. The accuracy rate must be higher than 97.5%.

https://victorzhou.com/blog/keras-neural-network-tutorial/

In [ ]:
#!pip install tensorflow==2.7.0 numpy mnist

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
     |████████████████████████████████| 665.5 MB 22 kB/s 
     |████████████████████████████████| 1.3 MB 35.0 MB/s 
     |████████████████████████████████| 463 kB 70.0 MB/s 
  Attempting uninstall: tensorflow-estimator
    Found existing installation: tensorflow-estimator 2.8.0
    Uninstalling tensorflow-estimator-2.8.0:
      Successfully uninstalled tensorflow-estimator-2.8.0
  Attempting uninstall: keras
    Found existing installation: keras 2.8.0
    Uninstalling keras-2.8.0:
      Successfully uninstalled keras-2.8.0
  Attempting uninstall: gast
    Found existing installation: gast 0.5.3
    Uninstalling gast-0.5.3:
      Successfully uninstalled gast-0.5.3
  Attempting uninstall: tensorflow
    Found existing installation: tensorflow 2.8.2+zzzcolab20220719082949
    Uninstalling tensorflow-2.8.2+zzzcolab20220719082949:
      Successfully uninstalled tensorflow-2.8.2+zzzcolab20220

In [ ]:
import numpy as np
import mnist
from tensorflow import keras

# The first time you run this might be a bit slow, since the
# mnist package has to download and cache the data.
train_images = mnist.train_images()
train_labels = mnist.train_labels()

print(train_images.shape) # (60000, 28, 28)
print(train_labels.shape) # (60000,)

(60000, 28, 28)
(60000,)


In [ ]:
import numpy as np
import mnist

train_images = mnist.train_images()
train_labels = mnist.train_labels()
test_images = mnist.test_images()
test_labels = mnist.test_labels()

# Normalize the images.
train_images = (train_images / 255) - 0.5
test_images = (test_images / 255) - 0.5

# Flatten the images.
train_images = train_images.reshape((-1, 784))
test_images = test_images.reshape((-1, 784))

print(train_images.shape) # (60000, 784)
print(test_images.shape)  # (10000, 784)

(60000, 784)
(10000, 784)


In [ ]:
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense, Dropout

model = Sequential([
Dense(64, activation='swish', input_shape=(784,)),
Dense(64, activation='sigmoid'),
Dense(10, activation='softmax'),
])

In [ ]:
from tensorflow.keras.optimizers import Adam
# Compile the model.
model.compile(
  optimizer='adam',
  loss='categorical_crossentropy',
  metrics=['accuracy'],
)

In [ ]:
from tensorflow.keras.utils import to_categorical
# Train the model.
model.fit(
  train_images,
  to_categorical(train_labels),
  epochs=10,
  batch_size=64,
  validation_data=(test_images, to_categorical(test_labels))
)

Epoch 1/10
938/938 [==============================] - 3s 3ms/step - loss: 0.4970 - accuracy: 0.8709 - val_loss: 0.2497 - val_accuracy: 0.9287
Epoch 2/10
938/938 [==============================] - 2s 3ms/step - loss: 0.2167 - accuracy: 0.9351 - val_loss: 0.1708 - val_accuracy: 0.9486
Epoch 3/10
938/938 [==============================] - 2s 2ms/step - loss: 0.1562 - accuracy: 0.9531 - val_loss: 0.1309 - val_accuracy: 0.9617
Epoch 4/10
938/938 [==============================] - 3s 3ms/step - loss: 0.1251 - accuracy: 0.9621 - val_loss: 0.1187 - val_accuracy: 0.9657
Epoch 5/10
938/938 [==============================] - 3s 3ms/step - loss: 0.1052 - accuracy: 0.9680 - val_loss: 0.1037 - val_accuracy: 0.9694
Epoch 6/10
938/938 [==============================] - 3s 3ms/step - loss: 0.0916 - accuracy: 0.9718 - val_loss: 0.1043 - val_accuracy: 0.9682
Epoch 7/10
938/938 [==============================] - 2s 3ms/step - loss: 0.0797 - accuracy: 0.9756 - val_loss: 0.0907 - val_accuracy: 0.9709
Epoch 

In [ ]:
model.evaluate(
  test_images,
  to_categorical(test_labels)
)

313/313 [==============================] - 0s 1ms/step - loss: 0.0789 - accuracy: 0.9758


[0.07888106256723404, 0.9757999777793884]

In [ ]:
model.save_weights('model.h5')

In [ ]:
# Predict on the first 5 test images.
predictions = model.predict(test_images[:5])

# Print our model's predictions.
print(np.argmax(predictions, axis=1)) # [7, 2, 1, 0, 4]

# Check our predictions against the ground truths.
print(test_labels[:5]) # [7, 2, 1, 0, 4]

[7 2 1 0 4]
[7 2 1 0 4]


  
> The initial accuracy rate is **96.87%**

1. First change the batch size and number of epochs

  Change batch size to 64 and epcho to 10
```
model.fit(
  train_images,
  to_categorical(train_labels),
  epochs=10,
  batch_size=64,
)
```

    The accuracy rate was **96.98%**, which is an improvement, but not by much.This proves that increasing epochs and batch size will have a slight improvement on the prediction accuracy


2. Change the number of hidden layers in the neural network
```
model = Sequential([
  Dense(64, activation='relu', input_shape=(784,)),
  Dense(64, activation='relu'),

  # New addition hidden layer
  Dense(64, activation='relu'),
  Dense(64, activation='relu'),
  
  Dense(10, activation='softmax'),
])
```

    Accuracy grew to **96.73%**, almost unchanged, even with a slight decrease. This can show that in this problem, changing the number of hidden layers has little effect on the final accuracy.

3. Change the activation function used
```
model = Sequential([
  Dense(64, activation='swish', input_shape=(784,)),
  Dense(64, activation='swish'),
  Dense(10, activation='softmax'),
])
```
    We change the activation function to swish. we can see that the final prediction accuracy is **96.93%**, which is a slight improvement. This proves that changing to a more optimal activation function is effective.

4. Change epchos, batch size and activation function at the same time.
```
model = Sequential([
Dense(64, activation='swish', input_shape=(784,)),
Dense(64, activation='sigmoid'),
Dense(10, activation='softmax'),
])
model.fit(
  train_images,
  to_categorical(train_labels),
  epochs=10,
  batch_size=64,
  validation_data=(test_images, to_categorical(test_labels))
)
```
```
313/313 [==============================] - 0s 1ms/step - loss: 0.0789 - accuracy: 0.9758
[0.07888106256723404, 0.9757999777793884]
```

 The accuracy was improved to **97.58%**, which can indicate that the combination of the two methods led to a significant improvement in the accuracy, and this method can be considered in the future when building neural network models.
  



